File: ts/step7-strategy-pattern/hardcodedRace.test.ts
import { Race, MovingStrategy } from './hardcodedRace';

describe('Race - Strategy 적용', () => {

    it('전략이 항상 true이면 모든 자동차가 이동한다', () => {
        const alwaysMove: MovingStrategy = () => true;

        const race = new Race(['kim', 'lee', 'park'], alwaysMove);
        race.playRound();

        const winners = race.findWinners();

        expect(winners).toEqual(
            expect.arrayContaining(['kim', 'lee', 'park'])
        );
        expect(winners).toHaveLength(3);
    });

    it('전략이 항상 false이면 아무도 이동하지 않는다', () => {
        const neverMove: MovingStrategy = () => false;

        const race = new Race(['kim', 'lee', 'park'], neverMove);
        race.playRound();

        const winners = race.findWinners();

        expect(winners).toEqual(
            expect.arrayContaining(['kim', 'lee', 'park'])
        );
        expect(winners).toHaveLength(3);
    });

    it('초기 상태에서는 모든 자동차가 우승자이다', () => {
        const dummyStrategy: MovingStrategy = () => false;

        const race = new Race(['kim', 'lee', 'park'], dummyStrategy);
        const winners = race.findWinners();

        expect(winners).toEqual(
            expect.arrayContaining(['kim', 'lee', 'park'])
        );
        expect(winners).toHaveLength(3);
    });

});

File: java/step7-strategy-pattern/src/test/java/badcode/HardcodedRaceTest.java
package badcode;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class HardcodedRaceTest {

    @DisplayName("playRound를 실행하면 자동차가 이동하지만 결과를 예측할 수 없다")
    @Test
    void playRound_isNotPredictable() {
        HardcodedRace race = new HardcodedRace(List.of("kim", "lee", "park"));

        // 😱 이 테스트는 랜덤에 의존하므로 결과를 보장할 수 없습니다!
        race.playRound();

        // 검증이 불가능 — 랜덤이라 어떤 결과가 나올지 모름
        // assertThat(???)

        // TODO: MovingStrategy 인터페이스를 도입하여,
        // 테스트에서는 고정된 값을 주입할 수 있도록 리팩토링하세요.
        // 예:
        // Race race = new Race(List.of("kim", "lee"), new FixedMovingStrategy(4));
        // race.playRound();
        // assertThat(race.findWinners()).containsExactlyInAnyOrder("kim", "lee");
    }

    @DisplayName("초기 상태에서는 모든 자동차가 우승자이다")
    @Test
    void findWinners_initialState() {
        HardcodedRace race = new HardcodedRace(List.of("kim", "lee", "park"));
        List<String> winners = race.findWinners();

        assertThat(winners).containsExactlyInAnyOrder("kim", "lee", "park");
    }
}
